src/handle.rs

```diff
@@ -419,3 +437,287 @@ impl HandleHost {
     unsafe { self.get_isolate().as_ref() }.thread_safe_handle()
   }
 }
+
+/// An object reference that does not prevent garbage collection for the object,
+/// and which allows installing finalization callbacks which will be called
+/// after the object has been GC'd.
+///
+/// Note that finalization callbacks are tied to the lifetime of a `Weak<T>`,
+/// and will not be called after the `Weak<T>` is dropped.
+///
+/// # `PartialEq` and `Hash`
+///
+/// The implementations of [`PartialEq`] and [`Hash`] might change during the
+/// lifetime of a `Weak<T>` instance: when compared via equals, a non-empty weak
+/// reference will not be equal to a different non-empty weak reference, but it
+/// will once both objects have been GC'd. Likewise, the hash of a weak
+/// reference will change after GC. That makes `Weak<T>` not suitable for use as
+/// the key for a [`HashMap`] or [`HashSet`].
+///
+/// # `Clone`
+///
+/// Since finalization callbacks are specific to a `Weak<T>` instance, cloning
+/// will create a new object reference without a finalizer, as if created by
+/// [`Self::new`]. You can use [`Self::clone_with_finalizer`] to attach a
+/// finalization callback to the clone.
+pub struct Weak<T> {
+  data: Option<Pin<Box<WeakData<T>>>>,
```

Comment:
Not super convinced this needs to be `Pin`.

Boxed values are implicitly pinned, except for the magical move-out-of-box operation (`let foo = *boxed_foo`). But since the `Box` is embedded in a struct here you wouldn't be able to do that anyway.

Reply:
That's not quite true, since `std::mem::swap` is a safe function:

```rust
// Self-referential struct
struct SelfRef {
  bytes: [u8; 1024],
  pointer: *const u8
}
impl SelfRef {
  pub fn new() -> Box<SelfRef> { todo!() }
  pub fn get_pointed_byte(&self) -> &u8 {
    // When the struct is created, `pointer` points into the `bytes` block, so this is fine, right?
    unsafe { &*self.pointer }
  }
}

let mut a = SelfRef::new();
let mut b = SelfRef::new();
std::mem::swap(&mut a, &mut b);  // !!!
drop(a);
b.get_pointed_byte();  // This is now a dangling reference
```

With `Pin<Box<T>>` you have access to `Pin<&mut T>`, but not to `&mut T` (except with unsafe methods or if `T: Unpin`), so you can't do `std::mem::swap`. You _can_ do [`Pin::set`](https://doc.rust-lang.org/std/pin/struct.Pin.html#method.set) because the Pin contract is that the value doesn't move for as long as it lives, and `Pin::set` drops the previous value.

In this case, we're not using `std::mem::swap` or anything like that, but since we're relying on the guarantee that the boxed data doesn't move, I guess it doesn't hurt to express that via the type system.

Reply:
Come to think about it, we don't really care that the value isn't moved, at least not for safety. We care that the box stays at that particular memory location and isn't deallocated until after the second pass callback, which is guaranteed by `Box` and by the fact that the box is never moved or deallocated until the weak handle is dropped, upon which `reset` is called, which means the callbacks won't be called.